sgeop/artifacts.py

```diff
@@ -22,29 +22,74 @@
-    isoareal_threshold_circles_enclosed=0.75,
-    isoperimetric_threshold_circles_touching=0.9,
-    exclusion_mask=None,
-    predicate="intersects",
-):
+    roads: gpd.GeoDataFrame,
+    threshold: None | float = None,
+    threshold_fallback: None | float = None,
+    area_threshold_blocks: float = 1e5,
+    isoareal_threshold_blocks: float = 0.5,
+    area_threshold_circles: float = 5e4,
+    isoareal_threshold_circles_enclosed: float = 0.75,
+    isoperimetric_threshold_circles_touching: float = 0.9,
+    exclusion_mask: None | gpd.GeoSeries = None,
+    predicate: str = "intersects",
+) -> tuple[gpd.GeoDataFrame, float]:
+    """Extract face artifacts and return the FAI threshold.
+    See :cite:`fleischmann2023` for more details.
+
+    Parameters
+    ----------
+    roads : geopandas.GeoDataFrame
+        Input roads that have been preprocessed.
+    threshold : None | float = None
+        First option threshold used to determine face artifacts. See the
+        ``artifact_threshold`` keyword argument in ``simplify.simplify_network()``.
+    threshold_fallback : None | float = None
+        Second option threshold used to determine face artifacts. See the
+        ``artifact_threshold_fallback`` keyword argument in
+        ``simplify.simplify_network()``.
+    area_threshold_blocks : float = 1e5
+        Areal theshold for block detection.
+    isoareal_threshold_blocks : float = 0.5
+        Isoareal theshold for block detection.
+        See ``esda.shape.isoareal_quotient``.
+    area_threshold_circles : float = 5e4
+        Areal theshold for circle detection.
+    isoareal_threshold_circles_enclosed : float = 0.75
+        Isoareal theshold for enclosed circle detection.
+        See ``esda.shape.isoareal_quotient``.
+    isoperimetric_threshold_circles_touching : float = 0.9
+        Isoperimetric theshold for enclosed circle touching.
+        See ``esda.shape.isoperimetric_quotient``.
+    exclusion_mask : None | gpd.GeoSeries = None
+        Polygons used to determine face artifacts to exclude from returned output.
+    predicate : str = 'intersects'
+        The spatial predicate used to exclude face artifacts from returned output.
+
+    Returns
+    -------
+    artifacts : geopandas.GeoDataFrame
+        Face artifact polygons.
+    threshold : float
+        Resultant artifact detection threshold from ``momepy.FaceArtifacts.threshold``.
+        May also be the returned value of ``threshold`` or ``threshold_fallback``.
+    """
+
+    def _relate(neighs: tuple, cond: callable) -> bool:
+        """Helper for relating artifacts."""
+        return len(neighs) > 0 and cond(polys.loc[list(neighs), "is_artifact"])
+
     with warnings.catch_warnings():  # the second loop likey won't find threshold
         warnings.filterwarnings("ignore", message="No threshold found")
         fas = momepy.FaceArtifacts(roads)
-    polygons = fas.polygons.set_crs(roads.crs)
+    polys = fas.polygons.set_crs(roads.crs)
```

Comment:
Out of curiosity, why?

Reply:
Shorter and used enough to make lines too long. TBH I like `polygons` better, but I was trying to limit lines of code. If you hate it we can change it back.

Reply:
I don't mind. I just wouldn't bother.